# HW2: Markov Models of Natural Language

*Fill in your name and the names of any students who helped you below.* 

> I affirm that I personally wrote the text, code, and comments in this homework assignment. 

> I did not recieve help on this assignment.

- *By Zoeb Jamal*
- Jan. 20th, 2021


## Language Models

Many of you may have encountered the output of machine learning models which, when "seeded" with a small amount of text, produce a larger corpus of text which is expected to be similar or relevant to the seed text. For example, there's been a lot of buzz about the new [GPT-3 model](https://en.wikipedia.org/wiki/GPT-3), related to its [carbon footprint](https://www.forbes.com/sites/robtoews/2020/06/17/deep-learnings-climate-change-problem/#2781c1b16b43), [bigoted tendencies](https://medium.com/fair-bytes/how-biased-is-gpt-3-5b2b91f1177), and, yes, impressive (and often [humorous](https://aiweirdness.com/)) [ability to replicate human-like text in response to prompts.](https://www.technologyreview.com/2020/07/20/1005454/openai-machine-learning-language-generator-gpt-3-nlp/) 

We are not going to program a complicated deep learning model, but we will construct a much simpler language model that performs a similar task. Using tools like iteration and dictionaries, we will create a family of **Markov language models** for generating text. For the purposes of this assignment, an $n$-th order Markov model is a function that constructs a string of text one letter at a time, using only knowledge of the most recent $n$ letters. You can think of it as a writer with a "memory" of $n$ letters. 

## Data

Our training text for this exercise comes from the first 10 chapters of Jane Austen's novel *Emma*, which I retrieved from the archives at [Project Gutenberg](https://www.gutenberg.org/files/158/158-h/158-h.htm#link2H_4_0001). Intuitively, we are going to write a program that "writes like Jane Austen," albeit in a very limited sense. 

In [25]:
s = "Today when I walked into my economics class I saw something I dread every time I close my eyes. Someone had brought their new gaming laptop to class. The Forklift he used to bring it was still running idle at the back. I started sweating as I sat down and gazed over at the 700lb beast that was his laptop. He had already reinforced his desk with steel support beams and was in the process of finding an outlet for a power cable thicker than Amy Schumer's thigh. I start shaking. I keep telling myself I'm going to be alright and that there's nothing to worry about. He somehow finds a fucking outlet. Tears are running down my cheeks as I send my last texts to my family saying I love them. The teacher starts the lecture, and the student turns his laptop on. The colored lights on his RGB Backlit keyboard flare to life like a nuclear flash, and a deep humming fills my ears and shakes my very soul. The entire city power grid goes dark. The classroom begins to shake as the massive fans begin to spin. In mere seconds my world has gone from vibrant life, to a dark, earth shattering void where my body is getting torn apart by the 150mph gale force winds and the 500 decibel groan of the cooling fans. As my body finally surrenders, I weep, as my school and my city go under. I fucking hate gaming laptops"

## Comments and Docstrings

This is the first homework assignment in which you will be graded in part on the quality of your documentation and explanation of your code. Here's what we expect: 

- **Comments**: Use comments liberally to explain the purpose of short snippets of code. 
- **Docstrings**: Functions (and, later, classes) should always be accompanied by a *docstring*. Briefly, the docstring should provide enough information that a user could correctly use your function ***without seeing the code.*** In somewhat more detail, the docstring should include the following information: 
    - One or more sentences describing the overall purpose of the function. 
    - An explanation of each of the inputs, including what they mean, their required data types, and any additional assumptions made about them.
    - An explanation of the output. 
    
In future homeworks, as well as on exams, we will be looking for clear and informative comments and docstrings. 

## Code Structure

In general, there are many good ways to solve a given problem. However, just getting the right result isn't enough to guarantee that your code is of high quality. Check the logic of your solutions to make sure that: 

- You aren't making any unnecessary steps, like creating variables you don't use. 
- You are effectively making use of the tools in the course, especially control flow. 
- Your code is readable. Each line is short (under 80 characters), and doesn't have long tangles of functions or `()` parentheses. 

Ok, let's go! 

## Exercise 1

Write a function called `count_characters()` that counts the number of times each character appears in a user-supplied string `s`. Your function should loop over each element of the string, and sequentually update a `dict` whose keys are characters and whose values are the number of occurrences seen so far. You may know of other ways to achieve the same result. However, you should use the loop approach, since this will generalize to the next exercise. 

*Note: while the construct `for letter in s:` will work for this exercise, it will not generalize to the next one. Use `for i in range(len(s)):` instead.* 

### Example usage: 

```python
count_characters("tortoise")
{'t' : 2, 'o' : 2, 'r' : 1, 'i' : 1, 's' : 1, 'e' : 1}
```

***Hint***: Yes, you did a problem very similar to this one on HW1. 


### Your Solution
To solve this problem, I will follow the same approach as in hw1, problem 2
- I will write a function to convert a string to a dictionary. I will create an empty `dict` in the body of the function and use a for loop to go through the elements of the string and write them in to the dictionary. In order to create the key-value pair, I will set each key equal to the number of times that element occurs in the original string. 

In [2]:
# write count_characters() here 
def count_characters(myString):
    """
    counts the number of times each character appears in a user-supplied string
    
    Parameters
    ----------
    myString = string that user wants to convert to dict
    
    Return
    ------
    d = dict whose keys are each unique character and corresponding value is the number of times it appears
    """
    d = {} # creating emtpy dict
    for i in range(len(myString)): # looping through the string
        d[myString[i]] = myString.count(myString[i])
        # each key of d is set to the i'th element of myString and its value is set to the number of times it occurs
        # I know that this will result in multiple occurences of each element of myString, but since dict keys are
        # dropped if they are repeated, this is not an issue
    return d

count_characters("tortoise")

{'t': 2, 'o': 2, 'r': 1, 'i': 1, 's': 1, 'e': 1}

## Exercise 2

An `n`-*gram* is a sequence of `n` letters. For example, `bol` and `old` are the two 3-grams that occur in the string `bold`. 

Write a function called `count_ngrams()` that counts the number of times each `n`-gram occurs in a string, with `n` specified by the user and with default value `n = 1`. You should be able to do this by making only a small modification to `count_characters()`. 

### Example usage: 

```python
count_ngrams("tortoise", n = 2)
```
```
{'to': 2, 'or': 1, 'rt': 1, 'oi': 1, 'is': 1, 'se': 1} # output
```

### Your Solution
To solve this problem, I will follow the same approach as in problem 1, with some minor adjustments
- I will need the keys of `d` to be `n`-grams instead of the `i`th element of the string. To do this, I can make a substring and set that as the `i`th key.
- Counting the number of times that `n`-gram appears will remain the same, but with the argument for the `.count()` method modified to reflect the substring
- I ran into a problem where I was getting keys in my dict that were not of the proper length. I think this had to do with how I was setting the keys of `d`. Essentially, the last `n - 1` keys of the dict would be of a shorter length than needed. For example, 
```python
count_ngrams("tortoise", n = 3)
```
```
{'tor': 1, 'ort': 1, 'rto': 1, 'toi': 1, 'ois': 1, 'ise': 1, 'se': 1, 'e': 1}
```
- To solve this problem, I added an if statement that checks if the substring we are about to create a key for is of the specified `n`-gram length. If it isn't, it will skip this value until `i` reaches the end of the string

In [3]:
# write count_ngrams() here
def count_ngrams(myString, n = 1):
    """
    counts the number of n-grams in a string
    
    Parameters
    ----------
    myString = string that the user enters
    n = the number of characters the user wants in each n-gram -> set to 1 by default
    
    Returns
    -------
    d = dict whose keys are the n-grams and values are the number of times they appear
    """
    d = {} # initialize empty dict
    for i in range(len(myString)): # loop through the string
        if len(myString[i : i + n]) == n: # need to make sure we are getting the correct length n-grams
            d[myString[i : i + n]] = myString.count(myString[i : i + n])
            # each key of d is set to a substring that contains n characters in myString -> value is number of occurences
            # I know that this will result in multiple occurences of each element of myString, but since dict keys are
            # dropped if they are repeated, this is not an issue
    return d

count_ngrams("tortoise", 2)

{'to': 2, 'or': 1, 'rt': 1, 'oi': 1, 'is': 1, 'se': 1}

## Exercise 3

Now we are going to use our `n`-grams to generate some fake text according to a Markov model. Here's how the Markov model of order `n` works: 

### A. Compute (`n`+1)-gram occurrence frequencies

You have already done this in Exercise 2!  

### B. Pick a starting (`n`+1)-gram

The starting (`n`+1)-gram can be selected at random, or the user can specify it. 

### C. Generate Text

Now we generate text one character at a time. To do so:

1. Look at the most recent `n` characters in our generated text. Say that `n = 3` and the 3 most recent character are `the`. 
2. We then look at our list of `n+1`-grams, and focus on grams whose first `n` characters match. Examples matching `the` include `them`, `the `, `thei`, and so on. 
3. We pick a random one of these `n+1`-grams, weighted according to its number of occurrences. 
4. The final character of this new `n+1` gram is our next letter. 

For example, if there are 3 occurrences of `them`, 4 occurrences of `the `, and 1 occurrences of `thei` in the n-gram dictionary, then our next character is `m` with probabiliy 3/8, `[space]` with probability 1/2, and `i` with probability `1/8`. 

**Remember**: the ***3rd***-order model requires you to compute ***4***-grams. 

## What you should do

Write a function that generates synthetic text according to an `n`-th order Markov model. It should have the following arguments: 

- `s`, the input string of real text. 
- `n`, the order of the model. 
- `length`, the size of the text to generate. Use a default value of 100. 
-  `seed`, the initial string that gets the Markov model started. I used `"Emma Woodhouse"` (the full name of the protagonist of the novel) as my `seed`, but any subset of `s` of length `n+1` or larger will work. 

Demonstrate the output of your function for a couple different choices of the order `n`. 


## Expected Output

Here are a few examples of the output of this function. Because of randomness, your results won't look exactly like this, but they should be qualitatively similar. 

```python
markov_text(s, n = 2, length = 200, seed = "Emma Woodhouse")
```
```
Emma Woodhouse ne goo thimser. John mile sawas amintrought will on I kink you kno but every sh inat he fing as sat buty aft from the it. She cousency ined, yount; ate nambery quirld diall yethery, yould hat earatte
```
```python
markov_text(s, n = 4, length = 200, seed = "Emma Woodhouse")
```

```
Emma Woodhouse!”—Emma, as love,            Kitty, only this person no infering ever, while, and tried very were no do be very friendly and into aid,    Man's me to loudness of Harriet's. Harriet belonger opinion an
```

```python
markov_text(s, n = 10, length = 200, seed = "Emma Woodhouse")
```

```
Emma Woodhouse's party could be acceptable to them, that if she ever were disposed to think of nothing but good. It will be an excellent charade remains, fit for any acquainted with the child was given up to them.
```

## Notes and Hints

***Hint***: A good function for performing the random choice is the `choices()` function in the `random` module. You can use it like this: 

```python
import random

options = ["One", "Two", "Three"]
weights = [1, 2, 3] # "Two" is twice as likely as "One", "Three" three times as likely. 

random.choices(options, weights) 
```

```
['One'] # output
```

The first and second arguments must be lists of equal length. Note also that the return value is a list -- if you want the value *in* the list, you need to get it out via indexing.  

***Hint***: The first thing your function should do is call `count_ngrams` above to generate the required dictionary. Then, handle the logic described above in the main loop.

In [14]:
# write markov_text() here
import random
def markov_text(s, n, length = 100, seed = "Emma Woodhouse"):
    """
    Generates fake text according to a Markov model. 
    
    Parameters
    ----------
    s = user supplied string
    n = order of the model
    length = the length of text to create (default = 100 characters)
    seed = the string to get the model to start
    
    Returns
    -------
    generated_text = the markov text created by this function
    """
    markov_dict = count_ngrams(s, n + 1) # creating a dict of ngrams for the input string
    markov_keys = list(markov_dict.keys()) # making a list of the keys and values in the dict
    markov_values = list(markov_dict.values())
    generated_text = seed # initializing the generated_text string to the seed
    for i in range(length): # need to loop for however many times the user requires
        recent = generated_text[-n:] # getting the  most recent n characters
        options = [] # making empty lists that will hold the ngrams and frequencies that match with recent
        weights = []
        for i in range(len(markov_keys)): # looping through all the ngrams
            if recent == markov_keys[i][0 : n]: # checking if recent matches with the first n characters of the ngrams
                options.append(markov_keys[i]) # adding the matches to the options list
                weights.append(markov_values[i]) # adding the frequences to the weights list
        
        random_choice = random.choices(options, weights) # getting a random ngram
        random_gram = random_choice[0] # converting random ngram to string
        generated_text += random_gram[-1] # adding the last character of random_gram to our generated_text
    return generated_text

In [31]:
# try out your function for a few different values of n
markov_text(s, n = 10, length = 1308, seed = "Today when I walked into ")

"Today when I walked into my economics class I saw something I dread every time I close my eyes. Someone had brought their new gaming laptop to class. The Forklift he used to bring it was still running idle at the back. I started sweating as I sat down and gazed over at the 700lb beast that was his laptop. He had already reinforced his desk with steel support beams and was in the process of finding an outlet for a power cable thicker than Amy Schumer's thigh. I start shaking. I keep telling myself I'm going to be alright and that there's nothing to worry about. He somehow finds a fucking outlet. Tears are running down my cheeks as I send my last texts to my family saying I love them. The teacher starts the lecture, and the student turns his laptop on. The colored lights on his RGB Backlit keyboard flare to life like a nuclear flash, and a deep humming fills my ears and shakes my very soul. The entire city power grid goes dark. The classroom begins to shake as the massive fans begin to 

In [23]:
# try another value of n! 
markov_text(s, n = 5, length = 200, seed = "Approximately")

'Approximately :o: 15 :b: people :man::woman: have died :skull: in HALF :skull: from COVID :mask::heart_eyes_cat::imp::japanese_ogre:. Not convinced :thinking:?Trump :tangerine::person_in_tuxedo_tone2::women_with_f'

In [24]:
# try third value! 
markov_text(s, n = 10, length = 200, seed = "Approximately")

'Approximately :o: 72813% more jobs :briefcase: in HALF :skull: of his :sweat_drops::eyes: FIRST :first_place: TERM :page_facing_up: than Biden :bangbang: has during in his :sweat_drops: glorious :innocent: reign :'

## Exercise 4

Using a `for`-loop, print the output of your function for `n` ranging from `1` to `10`. 

Then, write down a few observations. How does the generated text depend on `n`? How does the time required to generate the text depend on `n`? Do your best to explain each observation.  

In [78]:
# code here
for i in range(1, 10):
    print("n = " + str(i))
    %timeit markov_text(s, n = i, length = 200, seed = "Emma Woodhouse")
    print(markov_text(s, n = i, length = 200, seed = "Emma Woodhouse") + "\n")

n = 1
47.2 s ± 62.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Emma Woodhouse benght paf w tha, y, ncineriry ut ovet d Kn oeleld, grssh choorpun igr anthot is tt he atot whe bely nge mitid weeryoor o onditon r n; e itillirand bofuchaby haingr I owa visig buctllo h cersu ve:—ss

n = 2
40.1 s ± 5.55 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Emma Woodhousende.”Thing an way ing elphigir ther knintend was bery gones lon of Emme it frommay gend dead not he fave hat gings Isand thal hasty yed a re reany dearcesis itin al gire sookink wast itiond hertuntim 

n = 3
34.2 s ± 3.16 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Emma Woodhouse she him—norand which me of Miss only expectly othink I would happer on atter, will Fairy)—conce of propersual with an in olded as she world, claim this very sincliness mannot day, point.”“I subjectio

n = 4
31 s ± 2.15 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Emma Woodhouse's farth! the only obliged tempered receiving d

In [80]:
print("n = 10")
%timeit markov_text(s, n = 10, length = 200, seed = "Emma Woodhouse")
print(markov_text(s, n = 10, length = 200, seed = "Emma Woodhouse") + "\n")

n = 10
26.1 s ± 293 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Emma Woodhouse at last was off; but Mr. Knightley. “Success supposes endeavouring for the effect of shade, you know—in a joke—it is all a joke. We always speaks to the public favour; and she had no intellects of Hi



### Observations
- We can see that the "correctness"/complexity of the text increases significantly as n increases. While it is pretty much gibberish for lower n values, when you run the function with n =10, you actually get cohesive sentence structure (syntax and meaning may be lacking but at least spelling is correct). I think this is because the `n+1`-grams are higher (ie for n = 10 we would be using 11grams). In this case, the keys of the dict created by `count_ngrams()` will be much more reflective of what actual english looks like and therefore produce more "correct" words when we add new characters. 
- We can see that the time required goes down significantly as n increases from 1 to 10. The time required was 47.2 s ± 62.6 ms for n = 1, compared to 26.1 s ± 293 ms for n = 10. In cases where n is low, the `n+1`grams will also be really short. this leads to having more keys compared to the dict for `n+`grams where `n` is large. Since the dict for small `n` is much larger, it takes longer for the function to loop through the `markov_keys` list to see if our most recent `n` characters are a match. 
